### PR TITLE
Add --single-ended flag to iperf2 and iperf3  #456

### PR DESCRIPTION
**Pull request by vvidic**
Allows running throughput tests against servers running only
iperf2/iperf3 server on the default port.

### CODE REVIEW

**Comment by mfeit-internet2** [comment]
This needs to go through the process of incrementing the schema version for the test; failure to do that will result in confusing error messages when a task with the new switch is presented to a pScheduler that doesn't support it.

**Comment by vvidic** [comment]
On Thu, Mar 15, 2018 at 06:18:39PM +0000, Mark Feit wrote:
> This needs to go through the process of incrementing the schema
> version for the test; failure to do that will result in confusing
> error messages when a task with the new switch is presented to a
> pScheduler that doesn't support it.

Ok, but I don't know how to do that :)

I also wanted to ask if we should have --single-ended-port=1234
for cases where iperf server is not using the default port?

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by tonin** [comment]
@mfeit-internet2 For the schema version increment, is it only the `version` field in https://github.com/perfsonar/pscheduler/blob/master/pscheduler-test-throughput/throughput/enumerate ? Shall we increase it to `1.1`?

@vvidic Having an option like `--single-ended-port=1234` would be nice, yes.

**Comment by mfeit-internet2** [comment]
@tonin There's a bit more to it than that.  If things are added or removed from what would be a valid test spec, there has to be a `"schema": n` member, where `n` is a version number that gets incremented.  The default is `1`, which is where most of our stuff is.

I'll try and get something about how to do it written up and put in the pScheduler wiki this week.

**Comment by mfeit-internet2** [comment]
First cut of info on schema evolution:  https://github.com/perfsonar/pscheduler/wiki/Schema-Evolution

**Comment by vvidic** [comment]
Schema now updated too, so please review again.

**Comment by vvidic** [comment]
Let me know if we need to update anything here?

**Comment by mfeit-internet2** [comment]
If you've tested the changes and are comfortable with it going to production, I'll pull it.  My only hang-up was the schema numbering.

**Comment by vvidic** [comment]
On Thu, May 17, 2018 at 09:32:19AM -0700, Mark Feit wrote:
> If you've tested the changes and are comfortable with it going to
> production, I'll pull it.  My only hang-up was the schema numbering.

Yes, I think it should be good.  Maybe Antoine wants to give it
a quick spin before the merge?

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

File: pscheduler-test-throughput/tests/cli-to-spec_test.py [hunk 1]
```diff
@@ -27,6 +27,8 @@ def test_all_options(self):
             "congestion": {"val": "reno" },
             "client-cpu-affinity": {"val": "1" },
             "server-cpu-affinity": {"val": "0" },
+            "single-ended": {},
+            "single-ended-port": { "val": "5000" },
```
Comment (tonin) [inline pscheduler-test-throughput/tests/cli-to-spec_test.py]: Is there a reason to not use the regular 5201 iperf port?

**Comment by vvidic** [comment]
On Thu, May 17, 2018 at 10:10:25AM -0700, Antoine Delvaux wrote:
> If the single-ended-port is the port on the target iperf host, I think the default value should be the default iperf port (5201).
> 
> > @@ -27,6 +27,8 @@ def test_all_options(self):
>              "congestion": {"val": "reno" },
>              "client-cpu-affinity": {"val": "1" },
>              "server-cpu-affinity": {"val": "0" },
> +            "single-ended": {},
> +            "single-ended-port": { "val": "5000" },
> 
> Is there a reason to not use the regular 5201 iperf port?

I think this is just a test of some kind, so it should not matter
much what you put in here.

By default --single-ended uses the 5201 port and you only need
--single-ended-port=X if it is not on this default port.

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by tonin** [comment]
All is good then, merged!

**Pull request merged**